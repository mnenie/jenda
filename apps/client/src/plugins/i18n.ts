import { createI18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '@/shared/libs/vee-validate/rules'
import { customPluralRule } from '@/shared/libs/i18n/plurals'
import { getCurrentLocale } from '@/shared/libs/i18n/utils/getLocale'

interface MessageSchema {
  [key: string]: any
}

const ru_ru = import.meta.glob(
  '@/shared/libs/i18n/locales/**/ru-RU.ts',
  { eager: true },
)
const en_us = import.meta.glob(
  '@/shared/libs/i18n/locales/**/en-US.ts',
  { eager: true },
)
const zh_cn = import.meta.glob(
  '@/shared/libs/i18n/locales/**/zh-CN.ts',
  { eager: true },
)

function loadLang(modules: Record<string, any>) {
  const messages: { [key: string]: string } = {}

  Object.keys(modules).forEach((module) => {
    Object.assign(messages, { ...modules[module].default })
  })
  return messages
}

const messages = {
  'en-US': {
    ...loadLang(en_us),
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...loadLang(ru_ru),
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...loadLang(zh_cn),
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

export default i18n
