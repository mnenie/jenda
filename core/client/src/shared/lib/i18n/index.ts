import { useCookies } from '@vueuse/integrations/useCookies'
import { computed } from 'vue'
import { createI18n } from 'vue-i18n'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
  },
  'ru-RU': {
    ...ruLocale,
  },
  'zh-CN': {
    ...zhLocale,
  },
}
const cookies = useCookies()

const getCurrentLocale = computed(() => {
  const cookieLanguage = cookies.get('i18n')
  if (cookieLanguage) {
    return cookieLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
