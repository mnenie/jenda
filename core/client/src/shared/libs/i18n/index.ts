import { computed } from 'vue'
import { createI18n } from 'vue-i18n'

import en from '../vee-validate/rules/en'
import ru from '../vee-validate/rules/ru'
import zh from '../vee-validate/rules/zh'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...en,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ru,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zh,
    },
  },
}

const getCurrentLocale = computed(() => {
  const storageLanguage = localStorage.getItem('i18n')
  if (storageLanguage) {
    return storageLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
