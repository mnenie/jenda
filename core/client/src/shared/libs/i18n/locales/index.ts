export { default as enLocale } from './en-US'
export { default as ruLocale } from './ru-RU'
export { default as zhLocale } from './zh-CN'
