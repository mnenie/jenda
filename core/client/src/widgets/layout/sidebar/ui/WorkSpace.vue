<script setup lang="ts">
import { computed } from 'vue'
import { Icon } from '@iconify/vue'
import { useRoute } from 'vue-router/auto'
import type { WorkspaceLink } from '@/shared/config/types-shared'
import { useExpanded } from '@/shared/composables'
import { UiBadge, UiButton } from '@/shared/ui'

const props = defineProps<{
  links: WorkspaceLink[]
}>()

const route = useRoute()

function isCurrentPath(link: WorkspaceLink): boolean {
  if (route.name === 'boards') {
    return link.pathName === 'boards'
  }
  if (route.name === 'community') {
    return link.pathName === 'templates'
  }
  return route.path === `/${link.pathName}`
}

const pathName = computed(() => {
  return props.links.map(link => ({
    ...link,
    isActive: isCurrentPath(link),
  }))
})

const expanded = useExpanded()

const { isExpanded } = expanded.getExpanded()

const contentPosition = computed(() => {
  return isExpanded.value ? 'space-between' : 'center'
})
</script>

<template>
  <div class="flex flex-col space-y-1.5 justify-start mb-20px">
    <RouterLink
      v-for="link in pathName"
      :key="link.id"
      v-tooltip.right="{
        content: $t(`sidebar.${link.name}`),
        triggers: ['hover'],
        disabled: isExpanded,
      }"
      :to="`/${link.pathName}`"
      class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"
    >
      <UiButton
        :variant="link.isActive ? 'secondary' : 'ghost'"
        size="sm"
        class="w-full gap-2 shadow-none px-2 transition-all duration-200 ease"
        :class="link.isActive ? '!bg-neutral-200/40 dark:!bg-neutral-700/50' : ''"
        :style="{ padding: !isExpanded ? '0px' : '', justifyContent: contentPosition }"
      >
        <div class="flex items-center gap-2">
          <Icon :icon="link.icon" class="!w-15px !h-15px 2xl:(!w-4 !h-4) text-neutral-900 dark:text-neutral-200" />
          <span v-show="isExpanded" class="text-13px 2xl:text-sm !fw500 text-neutral-900 dark:text-neutral-100">{{ $t(`sidebar.${link.name}`) }}</span>
        </div>
        <UiBadge v-if="link.id === 1 && isExpanded" variant="solid" class="px-5px py-0 text-xs rounded-md">
          <span class="text-[10px] text-white">3</span>
        </UiBadge>
      </UiButton>
    </RouterLink>
  </div>
</template>
