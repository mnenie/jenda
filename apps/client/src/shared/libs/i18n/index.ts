import { createI18n, type I18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '../vee-validate/rules'
import { enLocale, ruLocale, zhLocale } from './locales'
import { customPluralRule } from './plurals'
import useDayJsLocalize from './dayjs'
import { getCurrentLocale } from './utils/getLocale'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

useDayJsLocalize(i18n as I18n)

export default i18n
