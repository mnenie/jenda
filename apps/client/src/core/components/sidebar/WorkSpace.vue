<script setup lang="ts">
import LinkItem from './links/LinkItem.vue'
import type { WorkspaceLink } from '../../types'
import { UiBadge } from '@/shared/ui'

defineProps<{
  links: WorkspaceLink[]
}>()
</script>

<template>
  <div class="flex flex-col space-y-1.5 justify-start mb-20px">
    <LinkItem
      v-for="link in links"
      :key="link.name"
      :link="link"
    >
      <template #badge="{ isExpanded }">
        <UiBadge
          v-if="link.badge && isExpanded && link.name === 'workflows'"
          variant="soft"
          class="px-5px py-0 text-xs rounded-md"
        >
          <span class="text-xs">beta</span>
        </UiBadge>
      </template>
    </LinkItem>
  </div>
</template>
