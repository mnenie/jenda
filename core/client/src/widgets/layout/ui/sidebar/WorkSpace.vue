<script setup lang="ts">
import type { Link } from '../../model'
import { UiButton } from '@/shared/ui'
import { Icon } from '@iconify/vue'
import { computed } from 'vue'
import { useRoute } from 'vue-router'

const props = defineProps<{
  links: Link[]
  isExpanded: boolean
}>()

const route = useRoute()

function isCurrentPath(link: Link) {
  if (route.path === '/boards') {
    return link.pathName === 'boards'
  }
  return route.path === `/${link.pathName}`
}

const pathName = computed(() => {
  return props.links.map(link => ({
    ...link,
    isActive: isCurrentPath(link),
  }))
})

const contentPosition = computed(() => {
  return props.isExpanded ? 'flex-start' : 'center'
})
</script>

<template>
  <div class="flex flex-col gap-5px justify-start mb-30px">
    <RouterLink
      v-for="link in pathName"
      :key="link.id"
      v-tooltip.right="{
        content: $t(`sidebar.${link.name}`),
        triggers: ['hover'],
        disabled: isExpanded,
      }"
      :to="link.pathName"
      class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"
    >
      <UiButton
        :variant="link.isActive ? 'secondary' : 'ghost'"
        class="w-full gap-2 shadow-none py-0 px-2 transition-all duration-200 ease"
        :class="link.isActive ? '!bg-neutral-200/40 dark:!bg-neutral-700/50' : ''"
        :style="{ padding: !isExpanded ? '0px' : '', justifyContent: contentPosition }"
      >
        <Icon :icon="link.icon" class="text-17px text-neutral-900 dark:text-neutral-200" />
        <span v-show="isExpanded" class="text-sm !fw500 text-neutral-900 dark:text-neutral-100">{{ $t(`sidebar.${link.name}`) }}</span>
      </UiButton>
    </RouterLink>
  </div>
</template>
