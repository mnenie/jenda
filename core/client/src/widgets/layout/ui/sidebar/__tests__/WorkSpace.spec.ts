import i18n from '@/shared/libs/i18n'
import { mount, RouterLinkStub } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { useRoute } from 'vue-router'
import WorkSpace from '../WorkSpace.vue'

vi.mock('vue-router')

vi.mock('@/shared/composables', () => ({
  useExpanded: vi.fn(() => ({
    getExpanded: vi.fn(() => ({
      isExpanded: {
        value: true,
      },
    })),
  })),
}))

describe('tests for WorkSpace.vue', () => {
  // @ts-expect-error mock types
  useRoute.mockReturnValue('/boards')
  const wrapper = mount(WorkSpace, {
    global: {
      plugins: [i18n],
      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            'sidebar.boards': 'Boards',
            'sidebar.section': 'Workspace',
          }
          return translations[key]
        },
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      directives: {
        tooltip() {},
      },
    },
    props: {
      links: [
        {
          id: 0,
          name: 'boards',
          pathName: 'boards',
          icon: 'i-hugeicons-trello',
        },
      ],
    },
  })

  it('should be render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be redirect with RouterLink correctly', () => {
    expect(wrapper.findComponent(RouterLinkStub).props('to')).toEqual('/boards')
  })
})
