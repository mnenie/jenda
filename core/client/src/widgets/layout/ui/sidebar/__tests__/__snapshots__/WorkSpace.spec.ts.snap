// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for WorkSpace.vue > should be render correctly 1`] = `
"<div class="flex flex-col gap-5px justify-start mb-20px"><a class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"><button class="btn btn-ghost h-34px w-full gap-2 shadow-none py-0 px-2 transition-all duration-200 ease" style="justify-content: space-between;">
      <!--v-if-->
      <div class="flex items-center gap-2"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" role="img" class="text-17px text-neutral-900 dark:text-neutral-200" width="1em" height="1em" viewBox="0 0 16 16"></svg><span class="text-sm !fw500 text-neutral-900 dark:text-neutral-100">Boards</span></div>
      <!--v-if-->
    </button></a></div>"
`;
